HW 32-33

fezhenko opened the pull request:
32 - change listener using spring;
33 - change project from servlets to spring MVC;

fezhenko committed:
creating model,repository,service for user requests

fezhenko committed:
create repository and service for requests, flyway migration

fezhenko committed:
Merge pull request #3 from fezhenko/main

main to hw_28

fezhenko committed:
add flyway pom

fezhenko committed:
create userRequest interface, repository, service, servlet, jsp page. Create databases for users, requests, friends.

fezhenko committed:
refactoring UserRequestRepository and related places

fezhenko committed:
refactoring filters, sql requests

fezhenko committed:
refactoring requests service,servlet,jsp

fezhenko committed:
refactoring requests service,servlet,jsp

fezhenko committed:
change DependencyInitializationContextListener using Spring

fezhenko committed:
adapt project for spring

fezhenko committed:
adapt project for spring mvc

andrey-kozel commented on lesson_26/src/main/java/org/example/login/filter/LoginFilter.java:
```diff
@@ -11,17 +11,26 @@
         final HttpServletRequest req = (HttpServletRequest) request;
         final HttpServletResponse res = (HttpServletResponse) response;
 
+        final List<String> filteredPages = new ArrayList<>();
+        filteredPages.add("users");
+        filteredPages.add("incoming-requests");
```
можно проинициализировать один раз в методе init

fezhenko replied:
private Set<String> availableUrls;
    @Override
    public void init(FilterConfig filterConfig) throws ServletException {
        availableUrls = new HashSet<>();
        availableUrls.add("/login");
        availableUrls.add("/registration");
        Filter.super.init(filterConfig);
    }

andrey-kozel commented on lesson_26/src/main/java/org/example/login/filter/LoginFilter.java:
```diff
@@ -11,17 +11,26 @@
+        filteredPages.add("users");
+        filteredPages.add("incoming-requests");
+
+        req.getServletPath();
```
это можно удалить)

fezhenko replied:
переписал использование filteredPages

andrey-kozel commented on lesson_26/src/main/java/org/example/login/filter/LoginFilter.java:
```diff
@@ -11,17 +11,26 @@
         final HttpServletRequest req = (HttpServletRequest) request;
         final HttpServletResponse res = (HttpServletResponse) response;
 
+        final List<String> filteredPages = new ArrayList<>();
```
а где эта переменная используется?

fezhenko replied:
переписал использование filteredPages

andrey-kozel commented on lesson_28/local-env/postgres/init.sql:
```diff
@@ -0,0 +1 @@
+CREATE DATABASE social_network;
```
это уже не нужно, база данных будет создана с помощью переменной "POSTGRES_DB"

fezhenko replied:
удалил init.sql

andrey-kozel commented on lesson_28/src/main/java/org/example/login/filter/SocialNetworkFilter.java:
```diff
@@ -0,0 +1,54 @@
+        final HttpServletRequest req = (HttpServletRequest) request;
+        final HttpServletResponse res = (HttpServletResponse) response;
+        final Set<String> filteredUrls = new HashSet<>();
+        filteredUrls.add("/users");
+        filteredUrls.add("/incoming-requests");
```
лучше сделать наоборот, записать сюда урлы страниц не требудющих авторизации) ведь выше вероятность, что будут добавляться урлы, которым авторизация нужна)

fezhenko replied:
отрефакторил, теперь задаются урлы к которым есть доступ в init методе фильтра

andrey-kozel commented on lesson_28/src/main/java/org/example/login/model/User.java:
```diff
@@ -0,0 +1,59 @@
+
+import java.util.Date;
+
+public class User {
```
можно использовать ломбок тут)

fezhenko replied:
отрефакторил, использую @Getter + @AllArgsConstructor

andrey-kozel commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Override
+    public void saveUser(User user) {
+        final String SAVE_USERNAME_AND_PASSWORD =
```
такой нэйминг используется для констант, просте переменные должны быть написаны в camelCase

andrey-kozel replied:
просмотрите другие места с этой проблемой тоже)

andrey-kozel commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,101 @@
+                        rs.getString("password"),
+                        rs.getDate("created_at"));
+            } else {
+                throw new IllegalArgumentException();
```
можно кинуть ошибку без else блока) он только добавляет лишние отступы) в случае ошибки тут лучше сделать кастомную UserNotFoundException)

fezhenko replied:
создал кастомный UserNotFoundException extend RuntimeException.
throw new UserNotFoundException("User with name:" + name + " is not found");

andrey-kozel commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRequestRepository.java:
```diff
@@ -0,0 +1,160 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class JdbcUserRequestRepository implements UserRequestRepository {
```
FriendRequest будет более удобным названием, описывающим то, что происходит) это ведь запрос на добавление в друзья)

fezhenko replied:
изменил на FriendRequest

Okhtienko commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRequestRepository.java:
```diff
@@ -0,0 +1,85 @@
+            ResultSet rs = statement.executeQuery(GET_ALL_USERS_REQUESTS);
+            final List<UserRequest> userRequests = new ArrayList<>();
+            while (rs.next()) {
+                final UserRequest userRequest = new UserRequest(
```
Создание userRequest лучше вынести в отдельные private метод и здесь вызывать этот метод.

andrey-kozel commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRequestRepository.java:
```diff
@@ -0,0 +1,160 @@
+        }
+    }
+
+    @Override
+    public List<User> findNotApprovedRequestsByUser(Long userId) {
+        final String GET_ALL_OUTCOMING_REQUESTS =
+                "SELECT u.user_id,u.name,u.role,created_at " +
+                "FROM requests r " +
+                "JOIN users u ON r.request_user_id =u.user_id " +
+                "WHERE r.received_user_id = ? AND isApproved = false";
+        try (PreparedStatement preparedStatement = connection.prepareStatement(GET_ALL_OUTCOMING_REQUESTS)) {
+            preparedStatement.setLong(1, userId);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            final List<User> requestedFriendUsers = new ArrayList<>();
+            while (resultSet.next()) {
+                final User requestedUser = new User(
+                        resultSet.getLong("user_id"),
+                        resultSet.getString("name"),
+                        resultSet.getString("role"),
+                        resultSet.getDate("created_at"));
+                requestedFriendUsers.add(requestedUser);
+            }
+            return requestedFriendUsers;
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
этот метод такой же, как и предыдущий

fezhenko replied:
удалил повторный метод

andrey-kozel commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRequestRepository.java:
```diff
@@ -0,0 +1,160 @@
+    @Override
+    public void createRequest(Long requestedUserId, Long receivedUserId) {
+        final String CREATE_REQUEST =
+                "INSERT INTO requests(request_user_id,received_user_id,isApproved)" +
```
в базе данных лучше называть поля в снэй кейсе, is_approved

fezhenko replied:
изменил на is_approved, исправил запросы

Okhtienko commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRequestRepository.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public List<UserRequest> findUserRequests() {
+        final String GET_ALL_USERS_REQUESTS = "select id,user_id,request_status,created_at from requests";
```
Запросы делать константами.

DenisSidorenkoQ commented on lesson_28/src/main/java/org/example/login/servlet/UsersServlet.java:
```diff
@@ -0,0 +1,31 @@
+@WebServlet("/users")
+public class UsersServlet extends HttpServlet {
+
+    private LoginService loginService;
```
Между полями и методами класса нужен отступ

andrey-kozel replied:
можно это настроить в чекстайле) часто комментарии, относящиеся к стилю закрываются именно таким образом)

andrey-kozel commented on lesson_28/src/main/webapp/users.jsp:
```diff
@@ -0,0 +1,63 @@
+    <c:if test="${param.error != null }">
+        <div class="row" style="margin-top: 25px">
+            <div class="alert alert-danger" role="alert">
+                    ${param.error}
```
используйте jstl)

Okhtienko commented on lesson_28/src/main/java/org/example/login/filter/UsersFilter.java:
```diff
@@ -0,0 +1,45 @@
+
+        final HttpSession session = req.getSession();
+        if (session.getAttribute("isLoggedIn") != null) {
+            final Boolean getIsLoggedInValue = (Boolean) req.getSession().getAttribute("isLoggedIn");
```
Лучше вынести переменную перед if.  Получается, что ты два раза получаешь значение, которое хранит сессия.

andrey-kozel commented on lesson_32/src/main/resources/db/migration/V2__Create_requests_table.sql:
```diff
@@ -0,0 +1,49 @@
+    FOREIGN KEY (request_user_id) REFERENCES users (user_id),
+    FOREIGN KEY (received_user_id) REFERENCES users (user_id)
+);
+-- (
+--     id BIGSERIAL NOT NULL UNIQUE,
+--     user_id BIGSERIAL NOT NULL,
+--     request_status VARCHAR NOT NULL CHECK (request_status in ('Incoming', 'Outcoming')),
+--     created_at TIMESTAMP NOT NULL DEFAULT now(),
+--     PRIMARY KEY (id),
+--     FOREIGN KEY (user_id) REFERENCES users (id)
+-- );
```
удалите закомментированный код в миграциях)

fezhenko replied:
удалил

Okhtienko commented on lesson_28/src/main/java/org/example/login/filter/UsersFilter.java:
```diff
@@ -0,0 +1,45 @@
+import javax.servlet.http.HttpSession;
+import java.io.IOException;
+
+@WebFilter(urlPatterns = "/users")
```
Фильтр работает только на один servlet. К остальным страницам неавторизованные пользователь имеет доступ.

DenisSidorenkoQ commented on lesson_32/src/main/java/org/example/socialnetwork/filter/SocialNetworkFilter.java:
```diff
@@ -0,0 +1,54 @@
+                res.sendRedirect("registration");
+            }
+            return;
+        }
```
3 if выглядят громоздки и их лучше избегать

andrey-kozel commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRequestRepository.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Override
+    public boolean isRequestApproved(Long requestedUserId, Long receivedUserId) {
```
вы передаете id в метод, а в запросе сравниваете по именам, по идее, так руботать не будет

fezhenko replied:
изменил запрос на использование user_id вместо имен

andrey-kozel commented on lesson_33/src/main/java/org/example/springmvc/config/JdbcConfig.java:
```diff
@@ -0,0 +1,28 @@
+            @Value("${username}") final String username,
+            @Value("${password}") final String password
+    ) throws SQLException {
+        System.out.println(driveClassName);
```
можно удалить)

fezhenko replied:
удалил

andrey-kozel commented on lesson_33/src/main/java/org/example/springmvc/controller/LoginController.java:
```diff
@@ -0,0 +1,37 @@
+
+    @GetMapping
+    protected String getLoginForm() {
+        return "login";
```
в принципе, так можно не делать, можно в контексте сделать страницу логина открытой создав view-controller)

Okhtienko commented on lesson_28/src/main/java/org/example/login/filter/UsersFilter.java:
```diff
@@ -0,0 +1,45 @@
+            final Boolean getIsLoggedInValue = (Boolean) req.getSession().getAttribute("isLoggedIn");
+            if (getIsLoggedInValue) {
+                chain.doFilter(request, response);
+                request.getServletContext().getRequestDispatcher("/users.jsp").forward(request, response);
```
Лишнее действие. Если getIsLoggedInValue, то отработает servlet и перенаправит на нужную страницу.

andrey-kozel commented on lesson_33/src/main/java/org/example/springmvc/controller/RegistrationController.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @PostMapping(consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    protected RedirectView createUser(final UserRegistrationDto userRegistrationDto) {
```
этот метод может быть частью UserController, по сути он просто создает пользователя)

fezhenko replied:
перенес метод в юзер контроллер в контроллере регистрации юзаю аннотацию @Autowired и использую метод из из юзер контроллера для создания юзера

Okhtienko commented on lesson_28/src/main/java/org/example/login/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,110 @@
+        User user = new User(name, password);
+        final String GET_NAME_PASSWORD_FROM_USERS_QUERY = "SELECT name,password FROM users WHERE name = ?" +
+                " and password = ?";
+        try {
```
Можно использовать try (PreparedStatement statement = connection.prepareStatement(GET_NAME_PASSWORD_FROM_USERS_QUERY)) {
....
} catch() {
} и тогда закрывать statement не нужно.

Okhtienko commented on lesson_33/src/main/java/org/example/springmvc/dto/UserLoginDto.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
```
лучше использовать аннотацию @Value.

Okhtienko commented on lesson_33/src/main/java/org/example/springmvc/model/User.java:
```diff
@@ -0,0 +1,59 @@
+
+import java.util.Date;
+
+public class User {
```
Можно использовать lombok.

Okhtienko commented on lesson_33/src/main/java/org/example/springmvc/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Override
+    public List<User> findUsers() {
+        final String GET_ALL_USERS = "select user_id,name,role,password,created_at from users";
```
Запросы выносить в константы за пределы метода.

Okhtienko commented on lesson_33/src/main/java/org/example/springmvc/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,96 @@
+            ResultSet rs = statement.executeQuery(GET_ALL_USERS);
+            final List<User> users = new ArrayList<>();
+            while (rs.next()) {
+                final User user = new User(
```
Создание пользователя выносить в отдельный private метод и здесь вызывать этот метод.

Okhtienko commented on lesson_33/src/main/java/org/example/springmvc/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,96 @@
+            ResultSet rs = statement.executeQuery();
+            if (rs.next()) {
+                return rs.getLong("user_id");
+            } else {
```
esle лишней. Передать исключение в catch и залогировать его там.

DenisSidorenkoQ commented on lesson_28/src/main/resources/db/migration/V2__Create_requests_table.sql:
```diff
@@ -0,0 +1,49 @@
+    FOREIGN KEY (request_user_id) REFERENCES users (user_id),
+    FOREIGN KEY (received_user_id) REFERENCES users (user_id)
+);
+-- (
```
Комментарии стоит убрать

DenisSidorenkoQ commented on lesson_33/src/main/java/org/example/springmvc/controller/RegistrationController.java:
```diff
@@ -0,0 +1,41 @@
+
+    @PostMapping(consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    protected RedirectView createUser(final UserRegistrationDto userRegistrationDto) {
+        if (userService.validateUser(userRegistrationDto.getName(), userRegistrationDto.getPassword())) {
```
Тут нужно делать валидацию по имени но не по паролю. У нас может быть юзер под таким же логином но с другим паролем и тогда он валидацию не пройдет и зарегистрируется под повторяющимся логином

fezhenko committed:
fix comments

fezhenko committed:
fix comments

fezhenko committed:
adjust project to use Spring Boot

fezhenko committed:
adjust project to use thymeleaf

fezhenko committed:
start to create REST controller

fezhenko committed:
add UsersRestController, implemented methods to get users, get user by id, create user.

fezhenko committed:
add Sprint security + JWT token

fezhenko committed:
add Sprint security + JWT token

fezhenko committed:
add encoded passwords to migration script
adjust user service
remove unnecessary requests, password encoder from JdbcUserRepository
fix api/auth, api/users endpoints

fezhenko committed:
create friend rest controller
creating FriendsRepository with JPA usage

fezhenko committed:
add RestController/Service/Repo to work with Friends

fezhenko committed:
add FriendRequestRepository/Service/RestController

fezhenko committed:
fix issues with RestControllers, implement new POST method to create UserRequest

fezhenko committed:
fix issues with RestControllers, implement new POST method to create UserRequest

andrey-kozel commented on .idea/encodings.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
файлы в папке .idea должны быть исключены в гите) они являются индивидуальными настройками для каждого разработчика и часто могут приводить к конфликтам)

andrey-kozel commented on lesson_26/src/main/java/org/example/login/filter/LoginFilter.java:
```diff
@@ -1,37 +1,56 @@
+        availableUrls = new HashSet<>();
+        availableUrls.add("/login");
+        availableUrls.add("/registration");
+        Filter.super.init(filterConfig);
```
это не обязательно, фильтр - это интерфейс без реализации) потому метод парента можно не вызывать)

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/controller/FriendRequestsController.java:
```diff
@@ -0,0 +1,67 @@
+    protected String approveFriendRequest(
+            @ModelAttribute("friendRequestsDto") @Valid final FriendRequestsDto friendRequestsDto, final Model model,
+            final BindingResult result) {
+        if (!result.hasErrors()) {
+            userRequestService.approveRequest(friendRequestsDto.getUserId(), authContext.getAuthUserId());
+        //TODO:сделать пост - при аппруе создаю в таблице френдс запись
+            final List<User> requestedUsers = userRequestService.findNotApprovedUserRequests(authContext.getAuthUserId());
+            model.addAttribute("requestedUsers", requestedUsers);
+            return "friendRequests";
+        }
```
friendRequestsDto тут не нужно добавлять как в гет методе? или это в гет методе оно лишнее?

fezhenko replied:
добавил в post метод так же

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/controller/FriendRequestsController.java:
```diff
@@ -0,0 +1,67 @@
+        return "friendRequests";
+    }
+
+    @PostMapping(consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE, params = "action=decline")
```
А что такое params?

fezhenko replied:
это контроллер понимает какую кнопку я нажал на странице. Ниже код для кнопки на странице таймлифа
`<button type="submit" name="action" value="accept" class="btn btn-success">Approve</button>`

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/controller/LoginController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @PostMapping(consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    protected String tryToLogin(
```
можно просто называть "login", по сути любой метод в джаве может вернуть ошибку и это не повод добавлять к ним приставку "tryTo" :)

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/controller/RegistrationController.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @PostMapping(consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    protected String getUserFromRegistrationPage(
```
имя метода не соответствует тому, что он делает

fezhenko replied:
change to userRegistration

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/controller/RegistrationController.java:
```diff
@@ -0,0 +1,45 @@
+@RequestMapping("/registration")
+@RequiredArgsConstructor
+public class RegistrationController {
+    UserController userController;
```
Контроллеры не должны ничего знать друг о друге

fezhenko replied:
перенес логику по созданию юзера в registrationsController

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/config/JdbcConfig.java:
```diff
@@ -0,0 +1,16 @@
+public class JdbcConfig {
+    @Bean
+    public Connection connection(JdbcTemplate jdbcTemplate) throws SQLException {
+        return jdbcTemplate.getDataSource().getConnection();
```
Вы где-то используете кнекшен напрямую? По идее, это можно удалить

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/repository/JdbcFriendRequest.java:
```diff
@@ -0,0 +1,121 @@
+import java.util.List;
+@Repository
+@RequiredArgsConstructor
+public class JdbcFriendRequest implements FriendRequest {
```
забыли в имена класса и интерфейса добавить имя слоя Repository

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/repository/JdbcFriendRequest.java:
```diff
@@ -0,0 +1,121 @@
+            "USING users u " +
+            "WHERE r.request_user_id = ? and r.received_user_id = ? and is_approved=false;";
+
+    private User createUser(Long userId, String name, String role, Date createdAt) {
+        return new User(userId, name, role, createdAt);
+    }
+    private List<User> getUsers(Long userId, String getUsersRequests) {
+        try (PreparedStatement preparedStatement = connection.prepareStatement(getUsersRequests,
+                Statement.RETURN_GENERATED_KEYS)) {
+            preparedStatement.setLong(1, userId);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            final List<User> users = new ArrayList<>();
+            while (resultSet.next()) {
+                users.add(createUser(
+                        resultSet.getLong("user_id"),
+                        resultSet.getString("name"),
+                        resultSet.getString("role"),
+                        resultSet.getDate("created_at"))
+                );
+            }
+            return users;
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
обычно приватные метод добавляются либо в конце класса, либо под методами, где они используются, такой подход сильно упрощает чтение кода, он просто читается сверху вниз)

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/repository/JdbcFriendRequest.java:
```diff
@@ -0,0 +1,121 @@
+            "JOIN users u ON r.request_user_id = u.user_id " +
+            "WHERE r.received_user_id = ? AND is_approved = false";
+    private static final String CREATE_REQUEST =
+            "INSERT INTO requests(request_user_id,received_user_id,is_approved)" +
+            "SELECT ?,?,false;";
```
обычно делается INSERT INTO ... (fields) VALUES (values), конструкция селект тоже подходит, но она обычно используется немного с другой целью)

fezhenko replied:
fixed

andrey-kozel commented on lesson_33/src/main/java/org/example/socialnetwork/service/FriendsService.java:
```diff
@@ -0,0 +1,4 @@
+package org.example.socialnetwork.service;
+
+public class FriendsService {
+}
```
А почему он пустой? Можно удалить, если ни где не используется)

fezhenko replied:
Просто не реализовал в этом модуле, удалил

andrey-kozel commented on lesson_33/src/main/resources/db/migration/V1__Create_users_table.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE IF NOT EXISTS appUsers
```
в именах таблиц обычно используется snake_case)

andrey-kozel commented on random/src/main/java/org/example/App.java:
```diff
@@ -0,0 +1,27 @@
+    {
+        System.out.println(getUser("name", "password"));
+    }
+    public static AppUser getUser(String name, String password) {
```
можно удалить этот модуль)

andrey-kozel commented on spring_security/local-env/postgres/init.sql:
```diff
@@ -0,0 +1 @@
+CREATE DATABASE social_network;
```
Это можно удалить, этот файл не используется в докер кампоузе)

andrey-kozel commented on spring_security/src/main/java/org/example/socialnetwork/controller/FriendRequestsController.java:
```diff
@@ -0,0 +1,65 @@
+package org.example.socialnetwork.controller;
+
+//import lombok.RequiredArgsConstructor;
```
Закоментированны код лучше просто удалять, он все-равно остается в истории гита

andrey-kozel commented on spring_security/src/main/java/org/example/socialnetwork/controller/FriendsRequestRestController.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/v1/friendRequests")
```
лучше использовать kebab-case для именования эндпоинтов

andrey-kozel commented on spring_security/src/main/java/org/example/socialnetwork/controller/FriendsRestController.java:
```diff
@@ -0,0 +1,62 @@
+        return ResponseEntity.ok(userConverter.toDto(friendsList));
+    }
+
+    @GetMapping("/{userId}/{friendId}")
```
api/v1/friends/{friendId}/users/{userId} - такой вариант больше соответствует REST спецификации

andrey-kozel commented on spring_security/src/main/java/org/example/socialnetwork/controller/FriendsRestController.java:
```diff
@@ -0,0 +1,62 @@
+        friendsService.removeFriend(userId, friendDto.getFriendId());
+    }
+
+    @PostMapping("/{userId}/approve")
+    @ResponseStatus(HttpStatus.OK)
+    public void approveFriend(
+            @PathVariable Long userId,
+            @RequestBody final FriendDto friendDto) {
+        friendsService.addFriend(userId, friendDto.getFriendId());
+    }
```
этот эндпоинт делает то же самое, что и adFriend, может, его можно удалить?

fezhenko replied:
удалил дубликат

andrey-kozel commented on spring_security/src/main/java/org/example/socialnetwork/controller/UsersRestController.java:
```diff
@@ -0,0 +1,58 @@
+
+    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
+    protected ResponseEntity<UserDto> createUser(@Valid @RequestBody final UserRegistrationDto userRegistrationDto) {
+        userService.createUser(userRegistrationDto.getUsername(), userRegistrationDto.getPassword(),
+                userRegistrationDto.getRole());
```
метод create ожет возвращать созданного юзера, чтобы не получать его из базы отдельным реквестом)

fezhenko replied:
изменил метод createUser  на уровне сервисов чтобы возвращать созданного юзера

andrey-kozel commented on spring_security/src/main/java/org/example/socialnetwork/repository/FriendRequestRepository.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.List;
+
+
+public interface FriendRequestRepository extends Repository<UserRequest, Long> {
```
может, этот репозиторий тоже перевести на JPA?) Хотя, это еще будет в следующих домашках)

andrey-kozel commented on spring_security/src/main/java/org/example/socialnetwork/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,124 @@
+//                .anyMatch(user -> user.getName().equals(username)
+//                        && user.getPassword().equals(password));
+//    }
+    @Override
+    public boolean validateUser(String username) {
+        return findUsers().stream()
+                .anyMatch(user -> user.getName().equals(username));
+    }
+
+    @Override
+    public AppUser getUser(String username) {
+        return findUsers().stream()
+                .filter(user -> user.getName().equals(username))
+                .findFirst()
+                .orElse(null);
+    }
+
+    @Override
+    public AppUser getUserById(Long userId) {
+        return findUsers().stream()
+                .filter(user -> user.getUserId().equals(userId))
+                .findFirst()
+                .orElse(null);
+    }
```
лучше это переписать на выборку из репозитория по параметрам и переписать на JPA)

fezhenko replied:
fixed

fezhenko committed:
fix comments

fezhenko committed:
init commit for new swagger module

fezhenko committed:
add swagger annotations to the rest UsersRestController

fezhenko committed:
create tests for UsersRestController

andrey-kozel commented on .idea/dataSources.xml:
```diff
@@ -15,5 +15,12 @@
       <jdbc-url>jdbc:postgresql://localhost:15432/postgres</jdbc-url>
       <working-dir>$ProjectFileDir$</working-dir>
     </data-source>
+    <data-source source="LOCAL" name="swagger@localhost" uuid="4a8a3589-8edd-4c29-84ea-651f07ee6a15">
```
Эту тоже не нужно комитить, добавьте в гит игнор .idea папку)

andrey-kozel commented on Swagger/Swagger.iml:
```diff
@@ -0,0 +1,82 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
и в целом *.iml файлы тоже нужно игноить)

andrey-kozel commented on Swagger/src/main/java/org/example/swagger/controller/FriendRequestsController.java:
```diff
@@ -0,0 +1,69 @@
+
+    @GetMapping
+    protected String getIncomingRequests(Model model) {
+        final List<UserRequest> requestedUsers = userRequestService.getUserRequests(1L); //TODO: find how to get userId
```
айди пользователя можно получить из секьюрити контекста)

andrey-kozel commented on Swagger/src/main/java/org/example/swagger/controller/FriendRequestsController.java:
```diff
@@ -0,0 +1,69 @@
+    protected String approveFriendRequest(
+            @ModelAttribute("friendRequestsDto") @Valid final FriendRequestsDto friendRequestsDto, final Model model,
+            final BindingResult result) {
+        if (!result.hasErrors()) {
+            userRequestService.approveRequest(friendRequestsDto.getRequestId());
+            //TODO:сделать пост - при аппруе создаю в таблице френдс запись
+            final List<UserRequest> requestedUsers = userRequestService.getUserRequests(1L);
+            model.addAttribute("requestedUsers", requestedUsers);
+            model.addAttribute("friendRequestsDto", new FriendRequestsDto());
+            return "friendRequests";
+        }
+        final List<UserRequest> requestedUsers = userRequestService.getUserRequests(1L);
+        model.addAttribute("requestedUsers", requestedUsers);
+        model.addAttribute("friendRequestsDto", new FriendRequestsDto());
+        return "friendRequests";
```
в Случае ошибок нужно сообить об ошибке, а то получается, что мы просто перерисовываем тот же самый юай)

andrey-kozel commented on Swagger/src/main/java/org/example/swagger/controller/FriendRequestsController.java:
```diff
@@ -0,0 +1,69 @@
+    protected String declineFriendRequest(
+            @ModelAttribute("friendRequestsDto") @Valid final FriendRequestsDto friendRequestsDto, final Model model,
+            final BindingResult result) {
+        if (!result.hasErrors()) {
+            userRequestService.declineRequest(friendRequestsDto.getRequestId());
+            final List<UserRequest> requestedUsers = userRequestService.getUserRequests(1L);
+            model.addAttribute("requestedUsers", requestedUsers);
+            model.addAttribute("friendRequestsDto", new FriendRequestsDto());
+            return "friendRequests";
+        }
+        final List<UserRequest> requestedUsers = userRequestService.getUserRequests(1L);
+        model.addAttribute("requestedUsers", requestedUsers);
+        model.addAttribute("friendRequestsDto", new FriendRequestsDto());
+        return "friendRequests";
```
Здесь тоже самое с сообщением об ошибке)

andrey-kozel commented on Swagger/src/main/java/org/example/swagger/controller/FriendsRequestRestController.java:
```diff
@@ -0,0 +1,93 @@
+        return ResponseEntity.ok(userRequestConverter.toDto(userRequest));
+    }
+
+    @PostMapping("/{requestId}/approve")
+    public ResponseEntity<String> approveFriendRequest(
+            @PathVariable Long requestId,
+            @RequestBody ApproveFriendRequestDto requestDto
+    ) {
+        userRequestService.approveRequest(requestId);
+        friendsService.addFriend(requestDto.getReceivedUserId(), requestId);
+        return ResponseEntity
+                .ok("user with id %d successfully approve user with id %d"
+                        .formatted(requestDto.getRequestedUserId(), requestDto.getReceivedUserId()));
+    }
+
+    @PostMapping("/{requestId}/decline")
+    public ResponseEntity<String> declineFriendRequest(
```
Когда мы делаем REST апи, все ответы должны быть в json формате, лучше тут тоже все обернуть в объект)

andrey-kozel commented on Swagger/src/main/java/org/example/swagger/controller/UsersRestController.java:
```diff
@@ -56,8 +56,17 @@ protected ResponseEntity<List<UserDto>> getUsers() {
     @GetMapping("/{userId}")
-    protected ResponseEntity<UserDto> getUser(@PathVariable Long userId) {
+    protected ResponseEntity<UserDto> getUser(@Valid @PathVariable final Long userId) {
+        if (userId == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(null);
+        }
```
можно поставить @NotBull на параметр метода

andrey-kozel commented on Swagger/src/main/java/org/example/swagger/dto/FriendRequestsDto.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.Date;
+
+@Data
```
почему над некоторыми Dto аннотация Data, а над некоторыми Value?

andrey-kozel commented on Swagger/src/main/java/org/example/swagger/dto/UserRegistrationDto.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.Data;
+import lombok.extern.jackson.Jacksonized;
+
+@Jacksonized
```
Jacksonized без билдера не нужен
